### DISCUSSION TITLE
How to get a list from a ListTable

### DISCUSSION THREAD

**Discussion by remeike** [circuithub/rel8#223] [Q&A] [answered]
So I've been trying to figure out how to use ListTables but can't seem to figure out how to work with the resulting data. I've been doing something similar to this example from the documentation:
```haskell
ordersWithItems :: Query (Order Expr, ListTable Expr (Item Expr))
ordersWithItems = do
   order <- each orderSchema
   items <- aggregate $ listAgg <$> itemsFromOrder order
   return (order, items)
```
My understanding is that if I run a select statement on this I should get back the following:
```haskell
resultingOrderWithItems :: Statement () [(Order Result, ListTable Result (Item Result))]
resultingOrderWithItems = select ordersWithItems
```
The trouble is I'm not sure what to do with `ListTable Result (Item Result)`. I was assuming I would be able to go from there to `[Item Result]`, but I haven't been able to figure it out. I suspect there's something more fundamental I'm missing here.

**Comment by ocharles** [comment, marked as answer]
Is that _actually_ what you get when you `select`? I'm pretty sure a `ListTable Expr t` should turn into `[t']`, where `t'` is whatever `t` would transpose to in the `Result` context.

**Reply by remeike** [reply]
Ok, I see where I had messed up here. I was kind of assuming that the result of the `select` would be `[(Order Result, ListTable Result (Item Result))]` so I hade a different function that was trying to handle that type but really it should have been written out as `[FromExprs (Order Result, ListTable Result (Item Result))]` or just `[(Order Result, [Item Result])]`. 

I think what really tripped me up is that originally I was trying to do this with a product type that wasn't a tuple. So really it looked more like this:
```haskell
data OrderItems f =
  OrderItems
    { order :: Order f
    , items :: ListTable f (Item f)
    }
  deriving stock Generic
  deriving anyclass Rel8able

ordersWithItems :: Query (OrderItems Expr)
ordersWithItems = do
   order <- each orderSchema
   items <- aggregate $ listAgg <$> itemsFromOrder order
   return (OrderItem order items)

resultingOrderWithItems :: Statement () [FromExprs (OrderItems Result)]
resultingOrderWithItems = select ordersWithItems
```
In this case the `items` field on `OrderItems` can't magically be turned into `[Item Result]` in the `Result` context, which I think is why I was trying to figure out if there was anything I could do with `ListTable Result (Item Result)`. So I suppose my question is really if there's a clever way to make this sort of thing work with records or if you're just better off sticking with tuples.

**Reply by ocharles** [reply]
> So I suppose my question is really if there's a clever way to make this sort of thing work with records or if you're just better off sticking with tuples.

Good news, there is! Try this:

```haskell
data OrderItems f =
  OrderItems
    { order :: Order f
    , items :: HList f (Item f)
    }
  deriving stock Generic
  deriving anyclass Rel8able
```

`HList` is a type family that will also change when `f` changes.

**Reply by remeike** [reply]
Brilliant! That works. Thank you!
